slurm_openstack_tools/reboot.py

```diff
@@ -1,160 +1,122 @@
+handler.setFormatter(logging.Formatter("%(asctime)s [%(levelname)s]: %(message)s"))
 logger.addHandler(handler)
 
-INSTANCE_UUID_FILE = "/var/lib/cloud/data/instance-id"
-
+# Directory containing per-node configurations
+HOSTVARS_DIR = "/exports/cluster/hostvars"
 
-def get_openstack_server_id():
-    if not path.exists(INSTANCE_UUID_FILE):
-        return None
-
-    with open(INSTANCE_UUID_FILE) as f:
-        return f.readline().strip()
-
-
-def get_sinfo_path():
-    # TODO(johngarbutt): get this from environment or config file?
-    sinfo_alt_path = "/usr/local/software/slurm/current/bin/sinfo"
-    if path.exists(sinfo_alt_path):
-        return sinfo_alt_path
-    return "sinfo"
-
-
-def get_reboot_reason():
-    # find our short hostname (without fqdn):
-    hostname = socket.gethostname().split(".")[0]
-    sinfo_path = get_sinfo_path()
-    # see why we're being rebooted:
-    sinfo = subprocess.run(
-        [
-            sinfo_path,
-            "--noheader",
-            "--nodes=%s" % hostname,
-            "-O",
-            "Reason:%i" % MAX_REASON_LENGTH,
-        ],
-        stdout=subprocess.PIPE,
-        universal_newlines=True,
-    )
-    return sinfo.stdout.strip()
-
-
-def get_image_from_reason(reason):
-    tokens = reason.split()
-    image = None
-    if len(tokens) > 1:
-        image_tokens = tokens[1].split(":")
-        if len(image_tokens) == 2 and image_tokens[0] == "image":
-            if image_tokens[1]:
-                image = image_tokens[1]
-                logger.info(f"user requested image:%{image}")
-    return image
-
-
-def rebuild_openstack_server(server_id, reason):
-    # Validate server_id
-    conn = openstack.connection.from_config()
-    server = conn.get_server(server_id)
 
-    image_descr = get_image_from_reason(reason)
-    if not image_descr:
-        image_descr = server.image.id
-        logger.info(
-            f"couldn't parse image from reason '%{reason}', falling back to "
-            f"existing image:%{image_descr}"
-        )
-
-    image = conn.image.find_image(image_descr)  # doesn't throw exception
-    if image is None:
-        logger.error(f"image {image_descr} either not found or not unique")
+def get_hostlist_from_slurm():
+    """
+    Parse the hostlist provided by Slurm as command-line arguments.
+    """
+    if len(sys.argv) < 2:
+        logger.error("No hostlist provided by Slurm")
         sys.exit(1)
-
-    # Note that OpenStack will power down the server as part of the rebuild
-    logger.info(f"rebuilding server %{server_id} with image %{image.id}")
-    conn.rebuild_server(server_id, image.id)
+    return sys.argv[1].split(",")
 
 
-def do_reboot():
-    # in case we are on an NFS share,
-    # try to ensure python stopped before reboot
-    os.execvp("reboot", ["reboot"])
+def read_hostvars(node):
+    """
+    Read the hostvars.yml file for a specific node.
+    """
+    hostvars_file = Path(HOSTVARS_DIR) / node / "hostvars.yml"
```

Comment:
Ultra picky here, lets validate that node only contains [A-Za-z0-9] or something like that, to avoid any shocks, security wise.